client/connectdlg_common.cpp

```diff
@@ -66,25 +56,53 @@
 #define WAIT_BETWEEN_TRIES 100000 /* usecs */
 #define NUMBER_OF_TRIES 500
 
-#if defined(HAVE_WORKING_FORK) && !defined(FREECIV_MSWINDOWS)
-/* We are yet to see FREECIV_MSWINDOWS setup where even HAVE_WORKING_FORK
- * would mean fork() that actually works for us. */
-#define HAVE_USABLE_FORK
-#endif
-
-#ifdef HAVE_USABLE_FORK
-static pid_t server_pid = -1;
-#elif FREECIV_MSWINDOWS
-HANDLE server_process = INVALID_HANDLE_VALUE;
-HANDLE loghandle = INVALID_HANDLE_VALUE;
-#endif /* HAVE_USABLE_FORK || FREECIV_MSWINDOWS */
 bool server_quitting = FALSE;
 
 static char challenge_fullname[MAX_LEN_PATH];
 static bool client_has_hack = FALSE;
 
 int internal_server_port;
 
+class serverProcess : public QProcess {
+  Q_DISABLE_COPY(serverProcess);
+
+public:
+  static serverProcess *i();
+
+private:
+  void drop();
+  serverProcess();
+  static serverProcess *m_instance;
+};
+serverProcess *serverProcess::m_instance = 0;
+
+/* Server process constructor */
+serverProcess::serverProcess()
+{
+  connect(this,
+          QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished),
+          [=](int exitCode, QProcess::ExitStatus exitStatus) {
+            qInfo() << _("Freeciv Server") << exitStatus;
+            drop();
+          });
+}
+/* Server process instance */
+serverProcess *serverProcess::i()
+{
+  if (!m_instance) {
+    m_instance = new serverProcess;
+  }
+  return m_instance;
+}
+/* Removes server process */
+void serverProcess::drop()
+{
+  if (m_instance) {
+    delete m_instance;
```

Comment:
Direct `delete` on a `QObject` is dangerous (there may be pending events still addressed at this object), use `deleteLater()`.